UnityProject/Assets/Scripts/Player/PlayerCrafting.cs

```diff
@@ -0,0 +1,350 @@
+	{
+		/// <summary>
+		/// The list of currently known recipes for a player by category.
+		/// </summary>
+		private List<List<CraftingRecipe>> knownRecipesByCategory = new List<List<CraftingRecipe>>();
+
+		[SerializeField, ReorderableList] [Tooltip("Default recipes known to a player.")]
+		private List<CraftingRecipe> defaultKnownRecipes = new List<CraftingRecipe>();
+
+		private PlayerScript playerScript;
+
+		public PlayerScript PlayerScript => playerScript;
+
+		private StandardProgressActionConfig craftProgressActionConfig = new StandardProgressActionConfig(
+			StandardProgressActionType.Craft
+		);
+
+		private void Awake()
+		{
+			playerScript = GetComponent<PlayerScript>();
+			InitKnownRecipesByCategories();
+			InitDefaultRecipes();
+		}
+
+		private void InitKnownRecipesByCategories()
+		{
+			for (int i = Enum.GetValues(typeof(RecipeCategory)).Length - 1; i >= 0; i--)
+			{
+				knownRecipesByCategory.Add(new List<CraftingRecipe>());
+			}
+		}
+
+		private void InitDefaultRecipes()
+		{
+			foreach (CraftingRecipe craftingRecipe in defaultKnownRecipes)
+			{
+				TryAddRecipeToKnownRecipes(craftingRecipe);
+			}
+		}
+
+		/// <summary>
+		/// 	Gets all the known recipes in the specified recipe category.
+		/// </summary>
+		/// <param name="recipeCategory">The recipe category.</param>
+		/// <returns>All the known recipes in the specified recipe category.</returns>
+		public List<CraftingRecipe> GetKnownRecipesInCategory(RecipeCategory recipeCategory)
+		{
+			return knownRecipesByCategory[(int) recipeCategory];
+		}
+
+		/// <summary>
+		/// 	Checks if a player already knows the recipe.
+		/// </summary>
+		/// <param name="recipe">The recipe to check.</param>
+		/// <returns>True, if a player already knows the recipe, false otherwise.</returns>
+		public bool KnowRecipe(CraftingRecipe recipe)
+		{
+			return GetKnownRecipesInCategory(recipe.Category).Contains(recipe);
+		}
+
+		/// <summary>
+		/// 	Adds the recipe to the KnownRecipesByCategory[] if it doesn't already contains the recipe.
+		/// </summary>
+		/// <param name="recipe">The recipe to add.</param>
+		public void LearnRecipe(CraftingRecipe recipe)
+		{
+			if (!TryAddRecipeToKnownRecipes(recipe))
+			{
+				return;
+			}
+
+			SendLearnedCraftingRecipe.SendTo(playerScript.connectedPlayer, recipe);
+		}
+
+		/// <summary>
+		/// 	Tries to add the recipe to the known recipes list.
+		/// </summary>
+		/// <param name="craftingRecipe">The recipe to add to the known recipes list.</param>
+		/// <returns>True if the recipe was successfully added to the known recipes list, false otherwise</returns>
+		public bool TryAddRecipeToKnownRecipes(CraftingRecipe craftingRecipe)
+		{
+			if (KnowRecipe(craftingRecipe))
+			{
+				return false;
+			}
+
+			UnsafelyAddRecipeToKnownRecipes(craftingRecipe);
+
+			return true;
+		}
+
+		/// <summary>
+		/// 	Adds the recipe to the known recipes list.
+		/// 	This method is unsafe - we may have duplicates!
```

Comment:
maybe could avoid duplicates using a HashSet instead of  a List

Reply:
I try to avoid using hash sets and dictionaries because they are greedy to memory.

I added this method to save a little bit of CPU time, when the programmer knows for sure that there will be no duplicate. A safer method is available above.

We will not have to add recipes often, so the O(n) complexity when adding recipes safely should not be felt much. But the memory sacrifices will be permanent, even if we'll have never used the safe method.